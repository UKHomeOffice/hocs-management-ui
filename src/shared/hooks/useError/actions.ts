import { FormError } from '../../models/formError';

export type AddFormError = {
    payload: FormError;
    type: 'AddFormError';
};

export type ClearError = {
    type: 'ClearError'
};

export type SetTitle = {
    payload: string;
    type: 'SetTitle';
};

export type SetDescription = {
    payload: string;
    type: 'SetDescription';
};

export type Action =
    AddFormError |
    ClearError |
    SetDescription |
    SetTitle;
