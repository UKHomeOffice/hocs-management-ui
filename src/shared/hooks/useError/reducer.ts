import { Action } from './actions';
import PageError from '../../models/pageError';

export const reducer = (state: PageError, action: Action): PageError => {
    switch (action.type) {
        case 'AddFormError':
            return {
                error: { ...state.error, formErrors: [...state.error ? state.error.formErrors || [] : [], action.payload] }
            };
        case 'ClearError':
            return { error: undefined };
        case 'SetDescription':
            return { error: { ...state.error, description: action.payload } };
        case 'SetTitle':
            return { error: { ...state.error, title: action.payload } };
    }
    return state;
};
