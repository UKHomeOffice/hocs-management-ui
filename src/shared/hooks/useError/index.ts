import React, { Reducer } from 'react';
import PageError from '../../models/pageError';
import { Action } from './actions';
import { reducer } from './reducer';
import { FormError } from '../../models/formError';
import ErrorMessage from 'shared/models/errorMessage';

const useError = (formErrorDescription?: string, formErrorTitle?: string)
    : [
        PageError,
        (formError: FormError) => void,
        () => void,
        (error: ErrorMessage) => void
    ] => {
    const [pageError, dispatch] = React.useReducer<Reducer<PageError, Action>>(reducer, {});

    const addFormError = (formError: FormError) => {
        (formErrorDescription || formErrorTitle) && setMessage({ title: formErrorTitle || '', description: formErrorDescription || '' });
        dispatch({ type: 'AddFormError', payload: formError });
    };
    const clearErrors = () => dispatch({ type: 'ClearError' });
    const setMessage = (errorMessage: ErrorMessage) => {
        dispatch({ type: 'SetDescription', payload: errorMessage.description });
        dispatch({ type: 'SetTitle', payload: errorMessage.title });
    };
    return [pageError, addFormError, clearErrors, setMessage];
};

export default useError;
