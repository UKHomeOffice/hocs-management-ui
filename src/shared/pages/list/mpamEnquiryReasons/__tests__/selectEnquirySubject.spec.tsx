import React from 'react';
import { match, MemoryRouter } from 'react-router-dom';
import { createBrowserHistory, History, Location } from 'history';
import { act, render, RenderResult, wait, fireEvent, waitForElement } from '@testing-library/react';
import Item from '../../../../models/item';
import * as useError from '../../../../hooks/useError';
import SelectEnquirySubject from '../selectEnquirySubject';

jest.mock('../../../../services/entityListService', () => ({
    __esModule: true,
    getListItems: () => Promise.resolve([{
        label: '__enquirySubject1__',
        value: '__enquirySubjectId1__'
    }, {
        label: '__enquirySubject2__',
        value: '__enquirySubjectId2__'
    }])
}));

let match: match<any>;
let history: History<any>;
let location: Location;
let wrapper: RenderResult;
let mockSelectedEnquirySubject: Item | undefined;

const useStateSpy = jest.spyOn(React, 'useState');
const setSelectedCaseTypeMock = jest.fn();
const useErrorSpy = jest.spyOn(useError, 'default');
const addFormErrorMock = jest.fn();
const clearErrorsMock = jest.fn();
const setMessageMock = jest.fn();
mockSelectedEnquirySubject = undefined;

const renderComponent = () => render(
    <MemoryRouter>
        <SelectEnquirySubject history={history} location={location} match={match}/>
    </MemoryRouter>
);

beforeEach(() => {
    history = createBrowserHistory();
    match = {
        isExact: true,
        params: { teamId: '__teamId__' },
        path: '',
        url: ''
    };

    location = {
        hash: '',
        key: '',
        pathname: '',
        search: '',
        state: {}
    };

    useStateSpy.mockImplementation(() => [mockSelectedEnquirySubject, setSelectedCaseTypeMock]);
    useErrorSpy.mockImplementation(() => [{}, addFormErrorMock, clearErrorsMock, setMessageMock]);
    history.push = jest.fn();
    addFormErrorMock.mockReset();
    clearErrorsMock.mockReset();
    setMessageMock.mockReset();
    act(() => {
        wrapper = renderComponent();
    });
});

describe('when the selectEnquirySubject component is mounted', () => {
    it('should render with default props', async () => {
        expect.assertions(1);

        await wait(() => {
            expect(wrapper.container).toMatchSnapshot();
            console.log(wrapper.container);
        });
    });
});

describe('when the submit button is clicked', () => {
    describe('and the data is filled in', () => {

        beforeAll(async () => {
            mockSelectedEnquirySubject = {
                label: '__enquirySubjectLabel__',
                value: '__enquirySubjectValue__'
            };
        });

        beforeEach(async () => {
            const submitButton = await waitForElement(async () => {
                return await wrapper.findByText('Submit');
            });

            fireEvent.click(submitButton);
        });

        it('should redirect to the home page', async () => {
            expect.assertions(1);

            await wait(() => {
                expect(history.push).toHaveBeenCalledWith('/enquiry-subject/__enquirySubjectValue__');
            });
        });
        it('should clear any previous errors', async () => {
            expect.assertions(1);

            await wait(() => {
                expect(clearErrorsMock).toHaveBeenCalled();
            });
        });
    });

    describe('and the data is not filled in', () => {

        beforeAll(async () => {
            mockSelectedEnquirySubject = undefined;
        });
        beforeEach(async () => {
            const submitButton = await waitForElement(async () => {
                return await wrapper.findByText('Submit');
            });

            fireEvent.click(submitButton);
        });

        it('should call the begin submit action', () => {
            expect(clearErrorsMock).toHaveBeenCalled();
        });

        it('should set the error state', () => {
            expect(addFormErrorMock).toHaveBeenNthCalledWith(1, { key: 'value', value: 'The Enquiry Subject is required' });
        });
    });
});
