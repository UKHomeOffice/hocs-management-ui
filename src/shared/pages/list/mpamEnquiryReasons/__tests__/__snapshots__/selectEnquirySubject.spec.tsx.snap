// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when the selectEnquirySubject component is mounted should render with default props 1`] = `
<div>
  <div
    class="govuk-grid-row"
  >
    <div
      class="govuk-grid-column-two-thirds-from-desktop"
    >
      <a
        class="govuk-back-link"
        href="/"
      >
        Back
      </a>
      <h1
        class="govuk-heading-xl"
      >
        Select an Enquiry Subject
      </h1>
    </div>
  </div>
  <div
    class="govuk-grid-row"
  >
    <div
      class="govuk-grid-column-one-half-from-desktop"
    >
      <form
        action="/api/template"
        method="POST"
      >
        <input
          name="_csrf"
          type="hidden"
          value=""
        />
        <div
          class="govuk-form-group"
        >
          <label
            class="govuk-label govuk-label--s"
            for="enquiry-subject-input"
            id="enquiry-subject-label"
          >
            Enquiry Subjects
          </label>
          <div
            class=" css-b62m3t-container"
            id="enquiry-subject"
          >
            <span
              class="css-1f43avz-a11yText-A11yText"
              id="react-select-2-live-region"
            />
            <span
              aria-atomic="false"
              aria-live="polite"
              aria-relevant="additions text"
              class="css-1f43avz-a11yText-A11yText"
              role="log"
            />
            <div
              class="govuk-typeahead__control css-1g8ily9-Control"
            >
              <div
                class="govuk-typeahead__value-container css-pilx6v-ValueContainer"
              >
                <div
                  class="govuk-typeahead__placeholder css-ovmm9z-Placeholder"
                  id="react-select-2-placeholder"
                >
                  Search
                </div>
                <div
                  class="govuk-typeahead__input-container css-qbdosj-Input"
                  data-value=""
                >
                  <input
                    aria-activedescendant=""
                    aria-autocomplete="list"
                    aria-describedby="react-select-2-placeholder"
                    aria-expanded="false"
                    aria-haspopup="true"
                    autocapitalize="none"
                    autocomplete="off"
                    autocorrect="off"
                    class="govuk-typeahead__input"
                    id="enquiry-subject-input"
                    role="combobox"
                    spellcheck="false"
                    style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                    tabindex="0"
                    type="text"
                    value=""
                  />
                </div>
              </div>
              <div
                class="govuk-typeahead__indicators css-1hb7zxy-IndicatorsContainer"
              >
                <span
                  class="govuk-typeahead__indicator-separator css-62wm9a-IndicatorSeparator"
                />
                <div
                  aria-hidden="true"
                  class="govuk-typeahead__indicator govuk-typeahead__dropdown-indicator css-hbt8bd-DropdownIndicator"
                >
                  <svg
                    aria-hidden="true"
                    class="css-tj5bde-Svg"
                    focusable="false"
                    height="20"
                    viewBox="0 0 20 20"
                    width="20"
                  >
                    <path
                      d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
        </div>
        <input
          class="govuk-button"
          type="submit"
          value="Submit"
        />
      </form>
    </div>
  </div>
</div>
`;
